proposals/safe-fixed-size-buffers.md

```diff
@@ -0,0 +1,558 @@
+
+In a member access of the form `E.I`, if member lookup of `I` identifies a non-readonly static fixed-size member,
+then `E.I` is evaluated and classified as a value of type `System.Span<S>`, where S is the element type of `I`.
+
+#### Readonly fixed-size buffers
+
+When a *field_declaration* includes a `readonly` modifier, the member introduced by the fixed_size_buffer_declarator is a ***readonly fixed-size buffer***.
+Direct assignments to elements of a readonly fixed-size buffer can only occur in an instance constructor, init member or static constructor in the same type.
+Specifically, direct assignments to an element of readonly fixed-size buffer are permitted only in the following contexts:
+
+- For an instance member, in the instance constructors or init member of the type that contains the member declaration; for a static member,
+  in the static constructor of the type that contains the member declaration. These are also the only contexts in which it is valid to pass
+  an element of readonly fixed-size buffer as an `out` or `ref` parameter.
+
+Attempting to assign to an element of a readonly fixed-size buffer or pass it as an `out` or `ref` parameter in any other context is a compile-time error.
+This is achieved by the following.
+
+A member access for a readonly fixed-size buffer is evaluated and classified as follows:
+
+- In a member access of the form `E.I`, if `E` is of a struct type and `E` is classified as a value, then `E.I` can be used only as a
+  *primary_no_array_creation_expression* of an [element access](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/expressions.md#11710-element-access)
+  with index of ```System.Index``` type, or of a type implicitly convertible to int. Result of the element access is a fixed-size member's element at the specified
+  position, classified as a value.
+- If access occurs in a context where direct assignments to an element of readonly fixed-size buffer are permitted, the result of the expression
+  is classified as a value of type `System.Span<S>`, where S is the element type of the fixed-size buffer. The value can be used to access member's elements.
+- Otherwise, the expression is classified as a value of type `System.ReadOnlySpan<S>`, where S is the element type of the fixed-size buffer.
+  The value can be used to access member's elements.
+
+### Definite assignment checking
+
+Fixed-size buffers are not subject to definite assignment-checking, and fixed-size buffer members are ignored for purposes of definite-assignment checking of struct type variables.
+
+When a fixed-size buffer member is static or the outermost containing struct variable of a fixed-size buffer member is a static variable, an instance variable of a class instance, or an array element, the elements of the fixed-size buffer are automatically initialized to their default values. In all other cases, the initial content of a fixed-size buffer is undefined.
```

Comment:
Does this is mean that the content will be uninitialized by default when the fixed-size buffer is a field of a local variable on stack? Is the program going to be able to see this uninitialized content without unsafe keyword?

Reply:
> Does this is mean that the content will be uninitialized by default when the fixed-size buffer is a field of a local variable on stack? Is the program going to be able to see this uninitialized content without unsafe keyword?

This proposal allows that, but we can require the buffers to be always initialized. Though, this is like to require additional changes in the language because this proposal doesn't support direct assignment to the safe fixed-size member in its entirety, and it might not be possible for compiler to verify that all elements in the member are definitely assigned (for example, when non-constant expressions are used to assign to elements).